src/main/scala/advent/solutions/Day3.scala

```diff
@@ -49,7 +53,172 @@ object Day3 {
     * @param text A comma separated string of displacements e.g "R75,D30,R83"
     */
   def parse(text: String): Either[WireParseError, Wire] = {
-    ???
+    val texts: List[(String, String)] =
+      text
+        .split(",")
+        .map(t => (t.filter(_.isLetter), t.filter(!_.isLetter)))
+        .toList
+
+    checkTexts(texts) match {
+      case "allValid" =>
+        Right(
+          Wire(texts.map(t => Displacement(whichDirection(t._1), t._2.toInt)))
+        )
+      case "invalidDirection" => Left(InvalidDirection(text))
+      case "invalidDistance"  => Left(InvalidDistance(text))
+    }
+  }
+
+  private def checkTexts(texts: List[(String, String)]): String = {
+    if (texts.forall(t => isValidDirection(t._1))) {
+      if (texts.forall(t => t._2.toInt > 0)) "allValid"
+      else "invalidDistance"
+    } else "invalidDirection"
+  }
+
+  private def isValidDirection(s: String): Boolean =
+    s == "L" | s == "R" | s == "U" | s == "D"
+
+  private def whichDirection(s: String): Direction = {
+    import Direction._
+    s match {
+      case "L" => Left
+      case "R" => Right
+      case "U" => Up
+      case "D" => Down
+    }
+  }
+
+  final case class Coordinate(x: Int, y: Int)
+
+  final case class StartingAndEndingPoints(start: Coordinate, end: Coordinate)
+
+  final case class TurningPoint(point: Coordinate)
+
+  final case class Path(points: Seq[Coordinate])
+
+  final case class WireMap(
+      displacement: Displacement,
+      startingAndEndingPoints: StartingAndEndingPoints
+  )
+
+  final case class PathLocations(displacement: Displacement, path: Path)
+
+  import Direction._
+
+  /**
+    *
+    * @param wire
+    * @return
+    */
+  def findTurningPoints(wire: Wire): List[TurningPoint] = {
+
+    @tailrec
+    def go(
+        accum: List[Coordinate],
+        currentCoordinate: Coordinate,
+        displacements: List[Displacement]
+    ): List[TurningPoint] = displacements match {
+      case x :: xs =>
```

Comment:
This would be a good point to start exploring `fold` :smiley: 
See if you can have a go rewriting the `go` functions to use `fold` instead of explicit recursion.